File: python-algorithm/ds_9_bit_operation/L231.py
# -*- coding: utf-8 -*-


# 若为2的幂，则转为二进制必定有且仅有一个1，其他全为0；

# n&(n-1)可以去除二进制中最右边的 1
class Solution:
    def isPowerOfTwo(self, n: int) -> bool:
        if n <= 0 :
            return False
        return (n&(n-1))==0

# n&（-n)可以得到二进制最右边的1
class Solution:
    def isPowerOfTwo(self, n: int) -> bool:
        if n == 0:
            return False
        return n & (-n) == n

File: ds_9_bit_operation/L231.cpp


//Bit Manipulation 78 136 169 231

//思路：利用位运算及二进制性质；
//若为2的幂，则转为二进制必定有且仅有一个1，其他全为0；
#include <iostream>

using namespace std;

class Solution {
public:
    bool isPowerOfTwo(int n) {
        if(n<=0) return false;
        return (n&(n-1))==0;
    }
};



